### ISSUE TITLE
[Bug] v0.5.14: With EAGLE/MTP on, KV pool profiler leaves ~50 GB VRAM idle on hybrid GDN model (Qwen3.6-27B NVFP4)

### ISSUE THREAD

**Issue by leehom0123** [sgl-project/sglang#29857]
## Bug description

On sglang **v0.5.14**, when EAGLE / MTP speculative decoding is enabled together with a hybrid GDN model, the profiler leaves a large amount of GPU memory idle and the KV pool token capacity is capped far below what the free VRAM can hold. Disabling speculative decoding lets the profiler fill VRAM as expected, so this is specific to the EAGLE-on path.

## Environment

- **Model**: `unsloth/Qwen3.6-27B-NVFP4` (Qwen3.6-27B, `Qwen3_5ForConditionalGeneration`, hybrid GDN + full-attention, NVFP4, compressed-tensors)
- **sglang version**: **0.5.14**
- **Docker image**: `lmsysorg/sglang:v0.5.14-cu130` (image id `43a6af27d491`)
- **GPU**: NVIDIA RTX PRO 6000 Blackwell, single GPU, 97887 MiB (~95.6 GiB) VRAM
- **Driver / CUDA**: 595.71.05 / CUDA 13.2 (host); CUDA 13.0.1 (image)
- **Host OS**: Ubuntu 24.04.4 LTS
- **kv_cache_dtype**: `fp8_e4m3`
- **context_length**: 262144
- **attention_backend**: flashinfer, **mamba_backend**: flashinfer
- **tp / pp / dp**: 1 / 1 / 1, **page_size**: 256

### Run 1 — EAGLE on, `mem_fraction_static=0.9`

```
--mem-fraction-static 0.9
--max-total-tokens 1572864
--max-running-requests 4
--chunked-prefill-size 32768
--context-length 262144
--kv-cache-dtype fp8_e4m3
--speculative-algorithm EAGLE
--speculative-num-steps 7
--speculative-eagle-topk 1
--speculative-num-draft-tokens 8
--mamba-full-memory-ratio 0.2   # default
--enable-hierarchical-cache
```

Log:

```
Load weight end. type=Qwen3_5ForConditionalGeneration, quant=compressed-tensors, avail mem=70.00 GB, mem usage=24.23 GB.
Load weight end. type=Qwen3_5ForCausalLMMTP,           quant=compressed-tensors, avail mem=64.43 GB, mem usage=5.53 GB.
max_total_tokens=1572864 is larger than the profiled value 322048. Use the profiled value instead.
Mamba Cache is allocated. max_mamba_cache_size: 48, conv_state size: 0.13GB, ssm_state size: 3.45GB intermediate_ssm_state_cache size: 2.81GB intermediate_conv_window_cache size: 0.05GB
KV Cache is allocated. dtype: torch.float8_e4m3fn, #tokens: 322048, K size: 4.92 GB, V size: 4.92 GB
Memory pool end. avail mem=48.11 GB
KV Cache is allocated. dtype: torch.float8_e4m3fn, #tokens: 322048, K size: 0.31 GB, V size: 0.31 GB
Memory pool end. avail mem=47.50 GB
max_total_num_tokens=322048, chunked_prefill_size=32768, max_prefill_tokens=32768, max_running_requests=4, context_len=262144, available_gpu_mem=50.97 GB
```

`nvidia-smi`: **46370 MiB used / 97887 MiB total** — only ~46 GB used, ~50 GB free after pool init. A 0.9 static budget should imply ~86 GB usage, but the profiler reports `available_gpu_mem=50.97 GB` free and stops there.

### Run 2 — same as Run 1 but EAGLE disabled (sanity check that the VRAM is actually available)

Remove `--speculative-*` args, set `--mamba-full-memory-ratio 0.9`, everything else identical.

```
max_total_tokens=1572864 is larger than the profiled value 1048320. Use the profiled value instead.
Mamba Cache is allocated. max_mamba_cache_size: 199, conv_state size: 0.55GB, ssm_state size: 28.12GB
KV Cache is allocated. dtype: torch.float8_e4m3fn, #tokens: 1048320, K size: 16.00 GB, V size: 16.00 GB
Memory pool end. avail mem=9.28 GB
max_total_num_tokens=1048320, chunked_prefill_size=8192, max_prefill_tokens=32768, max_running_requests=4, context_len=262144, available_gpu_mem=8.74 GB
```

`nvidia-smi`: **89570 MiB used** — with EAGLE off the profiler happily fills nearly the full VRAM. This confirms the ~50 GB left idle in Run 1 is not a hardware limit but a profiler decision made when EAGLE is on.

### Run 3 — current workable config (EAGLE off, tuned mamba budget)

```
--mem-fraction-static 0.77
--max-running-requests 4
--chunked-prefill-size 32768
--context-length 262144
--kv-cache-dtype fp8_e4m3
--mamba-full-memory-ratio 0.1
--enable-hierarchical-cache
--mm-attention-backend fa4
--enable-mixed-chunk
(no speculative decoding)
```

Log:

```
Load weight end. type=Qwen3_5ForConditionalGeneration, avail mem=70.00 GB, mem usage=24.23 GB.
Mamba Cache is allocated. max_mamba_cache_size: 60, conv_state size: 0.17GB, ssm_state size: 4.29GB
KV Cache is allocated. dtype: torch.float8_e4m3fn, #tokens: 1437952, K size: 21.95 GB, V size: 21.95 GB
Memory pool end. avail mem=21.56 GB
max_total_num_tokens=1437952, chunked_prefill_size=32768, max_prefill_tokens=32768, max_running_requests=4, context_len=262144, available_gpu_mem=21.02 GB
```

KV pool reaches **1,437,952 tokens** (~4.5× Run 1) at `mem_fraction_static=0.77`, using ~74 GB VRAM. Note this only works with EAGLE disabled.

### Comparison

| Run | mem_fraction | EAGLE / MTP | mamba_ratio | Profiled KV #tokens | Main KV (K+V) | Draft KV (K+V) | avail after pool |
|---|---:|:-:|---:|---:|---:|---:|---:|
| 1 | 0.9  | on  | 0.2 | **322,048**  | 9.84 GB  | 0.62 GB | 50.97 GB (⚠ ~50 GB idle) |
| 2 | 0.9  | off | 0.9 | 1,048,320    | 32.00 GB | —       | 8.74 GB |
| 3 | 0.77 | off | 0.1 | 1,437,952    | 43.90 GB | —       | 21.02 GB |

## Expected behavior

With EAGLE / MTP enabled, the KV pool profiler should still be able to fill up to the `--mem-fraction-static` budget instead of leaving ~50 GB of VRAM idle. Enabling speculative decoding should not, by itself, reduce KV pool capacity to ~30% of what the same GPU can hold.

## Additional context

- Reducing `--chunked-prefill-size` from 32768 → 8192 in the Run 1 config had zero effect on the profiled #tokens.
- Reducing `--mamba-full-memory-ratio` from 0.2 → 0.05 (with EAGLE still on) only bumped #tokens 322k → 346k and silently reduced `max_running_requests` from 4 to 2.
- Reproduced with `--enable-hierarchical-cache` both on and off.
- `available_gpu_mem=50.97 GB` in Run 1 is reported by sglang itself at the end of `Memory pool end`, confirming the profiler is aware that the memory is free but does not use it.

**Comment by waynehacking8** [comment]
Reproduced the arithmetic on the same card and model. On 0.5.14 your 322048 replays exactly: the draft rewrite sets num_nextn_predict_layers=1 on the outer hf_config, the sizing code reads the nested text config and gets None, falls back to the draft's full 64 layers, and the EAGLE cell-size scale becomes 1 + 64/16 = 5x.

On current main that path no longer engages for built-in MTP (no --speculative-draft-model-path): I booted your flag set with dummy weights, with and without a config fix, and both profile 1080576 tokens with VRAM filled (~10 GB free). So this looks specific to 0.5.14's sizing path - worth retesting on main or the next release. The outer-vs-nested config attr inconsistency itself is still present on main; I have a one-line fix with tests on a branch if the maintainers want it.

**Comment by robbiemu** [comment]
I repro this and extend the issue: the bug reproduces on the NVIDIA NVFP4 checkpoint (not just Unsloth's), and further `--speculative-attention-mode decode` does not work around it (the issue suggests hybrid-attention → decode; we tested it, pool still collapses).

@waynehacking8 verified locally, MTP is up locally for me with your solution on the NVIDIA variant that I am running of the 27b nvfp4 model (pool recovered 78K → 363K, VRAM honestly allocated). But I still have a bug I need to hunt down, I'm getting basically 0 acceptance. -- my PR mentioning this issue does not have that code, its for the NVIDIA NVFP4 model itself.

**Comment by waynehacking8** [comment]
Thanks for the retest -- that 78K -> 363K recovery on main settles the runtime question, so I've opened the fix as #30092 (fallback + regression tests). On the 0 acceptance: the first thing I'd check is whether the nextn/draft weights actually loaded under your modelopt_mixed routing -- your own PR notes the Qwen name-mapping mutates names across failed attempts, and silently unloaded draft weights produce exactly this signature (draft logits are noise, verify rejects everything). Dumping the norms of a couple of draft-layer tensors after load, or grepping the loader log for the mtp/nextn names, should tell you in a minute.

**Comment by waynehacking8** [comment]
Saw your question on #30078 about whether #30119 reaches the MTP case here -- from a code read it should: built-in MTP aliases the draft path to the main checkpoint (speculative_hook.py:346), so the draft ModelConfig runs through the exact override path #30119 patches (model_config.py:1343 lets the checkpoint's modelopt config overwrite whatever you set), which is why `--speculative-draft-model-quantization unquant` doesn't stick today. That makes a retest of your 0-acceptance with #30119 applied + `unquant` worthwhile -- nextn weights getting routed through modelopt_mixed is exactly the noise-logits signature from my last comment. Code read only, I haven't run it.

**Comment by FrySong** [comment]
There's a second, independent reason the pool comes out small on this model, separate from the cell-size scale.

The draft's `embed_tokens` and `lm_head` are still resident when `init_target_memory_pool()` profiles free memory. `init_lm_head()` frees them and repoints the draft at the target's tensors, but it runs at the end of `EagleDraftWorker.alloc_memory_pool()`, which the scheduler only reaches after the profile. At vocab 248320 / hidden 5120 each tensor is 2.54 GB, so the draft's copies are ~5 GB charged against the budget and then released into a number that is already fixed. Your log shows it too: `Qwen3_5ForCausalLMMTP ... mem usage=5.53 GB`.

Moving `init_token_map()` / `init_lm_head()` into `EagleDraftWorker.__init__` (plus an `empty_cache()`, since dropping the Parameter refs alone leaves the blocks in the caching allocator) gets it back. On H800 with dummy weights, `max_total_num_tokens` 258395 -> 336027, and a vocab sweep at 32k/128k/200k/248k has the recovered KV tracking 2*vocab*hidden*dtype to within 0.5%, so it is exactly those two tensors. Greedy output byte-identical.

PR: #32468

Additive to your fix, not a replacement. You mentioned ~10 GB still free on main after the config fallback fix - this could be part of it, though I have not run your flag set (page_size 256, fp8 KV, hierarchical cache) so I can't say how much of it.

**Comment by wudishidove** [comment]
Root cause found, on a different model and GPU but the same symptom. Posting the trace in case it saves someone
  the bisect.

  **Setup:** sglang 0.5.17, Qwen3.8-27B NVFP4 (hybrid GDN, 64 layers, `full_attention_interval=4` → 16
  full-attention layers, `mtp_num_hidden_layers: 1`), RTX 5090 32 GB, WSL2, NEXTN 4/1/5 with
  `--speculative-draft-model-path` pointing at the target checkpoint directory, `--kv-cache-dtype fp8_e4m3`,
  `--disable-radix-cache`, `--enable-linear-replayssm-spec`.

  ## The symptom is not a memory-accounting problem

  Freeing memory does not give the pool anything back:

  | change | memory freed | pool delta |
  |---|---|---|
  | `--max-running-requests` 4 → 2 | 0.28 GB of GDN state | **+437 tokens** |
  | override the 3712 MB pre-capture activation reserve to 1024 MB | 2.62 GB | **+2,593 tokens** |

  Five boots, fitting the KV budget against what the profiler had available:

  | avail after weights | mamba charge | pool | budget |
  |---|---|---|---|
  | 6.36 | 0.76 | 34,873 | 1.130 GiB |
  | 6.12 | 0.76 | 33,285 | 1.079 GiB |
  | 5.89 | 0.46 | 33,722 | 1.093 GiB |
  | 6.28 | 0.46 | 36,315 | 1.178 GiB |
  | 6.84 | 0.46 | 39,932 | 1.295 GiB |

  `budget = 0.200 * (avail - mamba)` fits all five within 0.3%.

  ## But the budget is correct — the KV cell size is 5x too large

  Instrumenting `_profile_available_bytes`, `_handle_max_mamba_cache` and `config_from_budget`:

  ```
  PROBE profile: pre_model_load=30.0908 mem_frac=0.99 avail=6.7836 slack=0.3009
                 post_capture=False pre_mamba=6.4827 post_mamba=6.0414
  PROBE mamba:   K=2 stage=0.1434GiB ring=0.0037GiB has_spec=True replayssm=True
                 state_mem=0.4414
  PROBE budget:  bytes=6486860677 (6.0414 GiB) cell=163840 -> tokens=39592
  ```

  The accounting is fine and generous: 6.04 GiB of budget, 0.30 GiB of slack, 0.44 GiB charged for mamba state. The
  problem is `cell=163840`.

  What actually gets allocated on that same run is `K size: 0.60 GB, V size: 0.60 GB` over 39,592 tokens = 32,768
  B/token (16 full-attention layers × 4 KV heads × 256 head_dim × 2 × 1 byte), plus 2,048 B/token of draft KV →
  **34,816 B/token**. The profiler prices every token at 163,840.

  `163840 / 32768 = 5.0` exactly. The chain:

  1. `model_executor/pool_configurator.py:130` — for a mambaish model `num_layers = len(effective_layer_ids)`, i.e.
  the **full-attention layers only** (16 of 64).
  2. `model_executor/pool_configurator.py:162` — `cell_size *= (1 + eagle_draft_num_layers / num_layers)`.
  3. `model_executor/model_runner_components/spec_aux_hidden_state.py:69-77` — `eagle_draft_num_layers =
  num_nextn_predict_layers` if set, otherwise `max(draft_model_config.num_hidden_layers,
  draft_model_config.num_attention_layers)`.

  For the Qwen3.5/3.6/3.8 family, `num_nextn_predict_layers` is `None` — that is the DeepSeek-style field name, and
  `ModelConfig` only sets it for an explicit list of architectures; these models declare their MTP depth as
  `mtp_num_hidden_layers` on the text config. And because `--speculative-draft-model-path` points at the target
  checkpoint, the fallback reads that model's own `num_hidden_layers` = **64**. So `1 + 64/16 = 5`, for an MTP head
  whose real KV cost is 1 layer (`1 + 1/16 = 1.0625`).

  There are arguably two bugs here:

  - the fallback returns the target's total layer count for a 1-layer MTP head; and
  - even with a correct draft count, the ratio puts a *total* layer count over a *full-attention-only* count. For a
  hybrid model those are different quantities, so the scale factor is wrong by `num_hidden_layers /
  num_full_attention_layers` whenever the draft count comes from the total.

  ## Local fix and result

  ```python
          num_nextn_predict_layers = draft_model_config.num_nextn_predict_layers
          if num_nextn_predict_layers is None:
              for _cfg in (
                  getattr(draft_model_config, "hf_text_config", None),
                  getattr(draft_model_config.hf_config, "text_config", None),
                  draft_model_config.hf_config,
              ):
                  _mtp = getattr(_cfg, "mtp_num_hidden_layers", None) if _cfg is not None else None
                  if _mtp:
                      num_nextn_predict_layers = int(_mtp)
                      break
          if num_nextn_predict_layers is not None:
              ...
  ```

  Same 32 GB card, same everything else: cell `163840 → 34816`, pool **33,285 → 104,953 tokens**,
  `--context-length` 32768 → 100000. A 96,451-token request (95k of text plus a 1600x900 image) now serves in 29.6
  s, and two concurrent 48k/49k-token sessions in 20.9 s.

  Decode throughput is unaffected (`bench.sh 512`, 3 reps per arm, same session, fix toggled by restoring the
  file):

  | | pool 40,274 | pool 104,953 |
  |---|---|---|
  | bs=1 | 128.6 / 119.6 / 129.4 → 125.9 tok/s | 126.9 / 134.9 / 123.3 → 128.4 tok/s |
  | bs=2 | 261.4 / 236.4 / 241.3 / 242.6 → 245.4 tok/s | 249.4 / 244.4 / 255.4 → 249.7 tok/s |

  **One consequence for whoever lands the fix:** with the cell corrected, the pool actually consumes the profiled
  budget, so `--mem-fraction-static` becomes load-bearing again. At 0.99 (which was harmless while the 5x cell
  capped the pool) the pool now takes everything and the first prefill OOMs. At 0.90 the same box leaves 3.01 GiB
  of slack, and 2 concurrent requests with 48,051 and 49,460 prompt tokens — one carrying a 1600x900 image, 97,511
  of 104,015 pool tokens live — peak at 31,677 MiB of 32,607, i.e. 930 MiB still free.

  **Suggested guard:** the profiler's `cell_size` and the pool's actual bytes-per-token at allocation are never
  cross-checked, so a 5x mismatch is completely silent — the only visible trace is a small pool and a lot of idle
  VRAM. A debug log or assert comparing the two at allocation time would have made this a one-boot diagnosis.

  For the original report: for a 64-layer model with `full_attention_interval=4` the factor should be exactly 5.
  The 3.25x measured there is across different `mem_fraction_static` values (0.9 with spec vs 0.77 without), so it
  is consistent in direction — logging the computed `cell_size` on that setup would confirm it is the same 5x.
